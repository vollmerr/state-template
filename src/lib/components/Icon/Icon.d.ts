import * as React from 'react';

import * as types from '../../utils/types';

export type IconName =
  | 'albums'
  | 'android'
  | 'apple'
  | 'apps'
  | 'archive'
  | 'arrow-down'
  | 'arrow-fill-down'
  | 'arrow-fill-left'
  | 'arrow-fill-left-down'
  | 'arrow-fill-left-up'
  | 'arrow-fill-right'
  | 'arrow-fill-right-down'
  | 'arrow-fill-right-up'
  | 'arrow-fill-up'
  | 'arrow-left'
  | 'arrow-next'
  | 'arrow-prev'
  | 'arrow-right'
  | 'arrow-up'
  | 'at-sign'
  | 'attachment'
  | 'audio'
  | 'bar-chart'
  | 'bear'
  | 'binoculars'
  | 'book'
  | 'brain'
  | 'briefcase'
  | 'building'
  | 'ca-capitol'
  | 'ca-state'
  | 'cal'
  | 'cal-bear'
  | 'calculator'
  | 'calendar'
  | 'call'
  | 'camera'
  | 'capitol'
  | 'caret-down'
  | 'caret-fill-down'
  | 'caret-fill-left'
  | 'caret-fill-right'
  | 'caret-fill-two-down'
  | 'caret-fill-two-left'
  | 'caret-fill-two-right'
  | 'caret-fill-two-up'
  | 'caret-fill-up'
  | 'caret-left'
  | 'caret-line-down'
  | 'caret-line-left'
  | 'caret-line-right'
  | 'caret-line-two-down'
  | 'caret-line-two-left'
  | 'caret-line-two-right'
  | 'caret-line-two-up'
  | 'caret-line-up'
  | 'caret-right'
  | 'caret-two-down'
  | 'caret-two-left'
  | 'caret-two-right'
  | 'caret-two-up'
  | 'caret-up'
  | 'carousel-next'
  | 'carousel-pause'
  | 'carousel-play'
  | 'carousel-prev'
  | 'cc-copyright'
  | 'certificate'
  | 'certificate-check'
  | 'charge'
  | 'charge-cycle'
  | 'charge-units'
  | 'chat'
  | 'chat-bubble'
  | 'chat-bubbles'
  | 'check-fill'
  | 'check-line'
  | 'check-list'
  | 'check-mark'
  | 'city'
  | 'clipboard'
  | 'clock'
  | 'close-fill'
  | 'close-line'
  | 'close-mark'
  | 'cloud-download'
  | 'cloud-gear'
  | 'cloud-services'
  | 'cloud-sync'
  | 'cloud-upload'
  | 'code'
  | 'collapse'
  | 'compass'
  | 'computer'
  | 'construction'
  | 'contact-us'
  | 'contacts'
  | 'currency'
  | 'document'
  | 'download'
  | 'drawer'
  | 'drive'
  | 'ear'
  | 'ear-slash'
  | 'email'
  | 'excel'
  | 'expand'
  | 'eye'
  | 'eye-slash'
  | 'facebook'
  | 'favorite'
  | 'file'
  | 'file-audio'
  | 'file-certificate'
  | 'file-check'
  | 'file-code'
  | 'file-csv'
  | 'file-download'
  | 'file-excel'
  | 'file-export'
  | 'file-import'
  | 'file-invoice'
  | 'file-medical'
  | 'file-medical-alt'
  | 'file-pdf'
  | 'file-powerpoint'
  | 'file-prescription'
  | 'file-upload'
  | 'file-video'
  | 'file-word'
  | 'file-zip'
  | 'film'
  | 'filter'
  | 'filter-solid'
  | 'fingerprint'
  | 'fingerprint-check'
  | 'fire'
  | 'flickr'
  | 'flowchart'
  | 'folder'
  | 'gear'
  | 'gears'
  | 'globe'
  | 'google-plus'
  | 'graduate'
  | 'graph'
  | 'green'
  | 'grid'
  | 'hand'
  | 'hand-money'
  | 'handshake'
  | 'home'
  | 'hourglass'
  | 'hours'
  | 'hours-security'
  | 'house'
  | 'idea'
  | 'image'
  | 'images'
  | 'important'
  | 'important-line'
  | 'info'
  | 'info-bubble'
  | 'info-line'
  | 'instagram'
  | 'institute'
  | 'justice-legal'
  | 'key'
  | 'law-enforcement'
  | 'lightbulb'
  | 'link'
  | 'linkedin'
  | 'location'
  | 'lock'
  | 'logo'
  | 'loud-speaker'
  | 'magnify-glass'
  | 'mail'
  | 'map'
  | 'medical'
  | 'medical-bubble'
  | 'medical-care'
  | 'medical-case'
  | 'medical-clinic'
  | 'medical-cross'
  | 'medical-doctor'
  | 'medical-heart'
  | 'medical-pills'
  | 'menu'
  | 'menu-toggle-closed'
  | 'menu-toggle-open'
  | 'mic'
  | 'microsoft'
  | 'minus-fill'
  | 'minus-line'
  | 'minus-mark'
  | 'mobile'
  | 'music'
  | 'online-services'
  | 'page'
  | 'pause'
  | 'payment-card'
  | 'pdf'
  | 'pencil'
  | 'pencil-edit'
  | 'people'
  | 'person'
  | 'phone'
  | 'pie-chart'
  | 'pinterest'
  | 'play'
  | 'plus-fill'
  | 'plus-line'
  | 'plus-mark'
  | 'powerpoint'
  | 'print'
  | 'pro-services'
  | 'pushpin'
  | 'puzzle'
  | 'puzzle-piece'
  | 'question'
  | 'question-fill'
  | 'question-line'
  | 'quote-fill'
  | 'read-book'
  | 'recycle'
  | 'responsive'
  | 'responsive-alt'
  | 'road-pin'
  | 'roadways'
  | 'rss'
  | 'science'
  | 'search'
  | 'search-right'
  | 'searching'
  | 'security-network'
  | 'security-system'
  | 'server'
  | 'share'
  | 'share-button'
  | 'share-email'
  | 'share-facebook'
  | 'share-flickr'
  | 'share-googleplus'
  | 'share-instagram'
  | 'share-linkedin'
  | 'share-pinterest'
  | 'share-twitter'
  | 'share-vimeo'
  | 'share-youtube'
  | 'shield'
  | 'shield-check'
  | 'shopping-cart'
  | 'skip-backwards'
  | 'skip-forward'
  | 'smartphone'
  | 'sos'
  | 'state'
  | 'table'
  | 'tablet'
  | 'tags'
  | 'thumb-up'
  | 'time'
  | 'tool'
  | 'tools'
  | 'travel-air'
  | 'travel-car'
  | 'triangle-down'
  | 'triangle-left'
  | 'triangle-line-down'
  | 'triangle-line-left'
  | 'triangle-line-right'
  | 'triangle-line-up'
  | 'triangle-right'
  | 'triangle-up'
  | 'trophy'
  | 'truck-delivery'
  | 'twitter'
  | 'user-id'
  | 'users'
  | 'users-alt'
  | 'users-dialog'
  | 'users-interaction'
  | 'video'
  | 'video-camera'
  | 'vimeo'
  | 'volume'
  | 'wallet'
  | 'warning-fill'
  | 'warning-triangle'
  | 'word'
  | 'youtube'
  | 'zipped-file'
  | 'zoom-in'
  | 'zoom-out'
;

export interface IconProps {
  /** Name of icon to use (auto prefixes with ca-gov-icon-) */
  name: IconName;

  /** Class names to attach to the top level div */
  className?: string;

  /** Screen Reader only text */
  srOnly?: string;

  /** HTML tag to render as  */
  tag?: React.ReactNode;
}

declare class Icon extends React.Component<IconProps, {}> {}

export default Icon;
