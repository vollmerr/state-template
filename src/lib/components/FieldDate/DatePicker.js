import React, { Component } from 'react';
import T from 'prop-types';
import Pikaday from 'pikaday';

import { isValid } from '../../utils/date';
import Icon from '../Icon';

// Date picker that uses pikaday...
class DatePicker extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      displayText: '',
    };
    this.pickerRef = React.createRef();
    this.displayRef = React.createRef();
    this.picker = null;
  }

  componentDidMount() {
    this.initPikaday();
  }

  componentWillReceiveProps(nextProps) {
    const { input } = this.props;
    const newValue = nextProps.input.value;
    if (input.value !== newValue) {
      // update displayed text when value changes
      const displayText = this.getDate(newValue).slice(0, 10);
      this.setState({ displayText });
      // update date in picker
      this.picker.setDate(newValue);
    }
  }

  getDate = (date) => {
    if (isValid(date)) {
      return new Date(date).toISOString();
    }
    return '';
  }

  onSelect = (date) => {
    const { input } = this.props;
    const newDate = this.getDate(date);
    // update redux-form
    input.onChange(newDate);
    // update display field
    this.setState({ displayText: newDate.slice(0, 10) });
  }

  initPikaday = () => {
    const { input, minDate } = this.props;

    this.picker = new Pikaday({
      field: this.pickerRef.current,
      onSelect: this.onSelect,
      minDate,
      yearRange: [new Date().getFullYear(), new Date().getFullYear() + 30],
      trigger: this.displayRef.current,
    });

    this.picker.setDate(input.value);
  }

  render() {
    const {
      input, disabled, minDate, ...rest
    } = this.props;
    const { displayText } = this.state;

    return (
      <div className="field-date has-feedback">
        {/* field that will be displayed */}
        <input
          ref={this.displayRef}
          className={'form-control'}
          disabled={disabled}
          value={displayText}
          onChange={() => {}}
          name={input.name}
          autoComplete={'off'}
        />
        {
          !disabled
          && <Icon name={'calendar'} className={'form-control-feedback'} />
        }
        {/* date picker */}
        <input
          ref={this.pickerRef}
          {...rest}
          className={'hidden'}
        />
      </div>
    );
  }
}

DatePicker.propsTypes = {
  /** redux-form input object */
  input: T.object.isRequired,
  /** picker is disabled */
  disabled: T.bool,
  /** min date to select from */
  minDate: T.oneOfType([T.string, T.instanceOf(Date)]),
};

DatePicker.defaultProps = {
  disabled: false,
  minDate: null,
};

export default DatePicker;
