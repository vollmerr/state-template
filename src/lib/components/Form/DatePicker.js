import React, { Component } from 'react';
import Pikaday from 'pikaday';

class DatePicker extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      displayText: '',
    };
    this.pickerRef = React.createRef();
    this.displayRef = React.createRef();
    this.picker = null;
  }

  componentDidMount() {
    this.initPikaday();
  }

  componentWillReceiveProps(nextProps) {
    // handle form resets
    if (!nextProps.input.value) {
      this.setState({ displayText: '' });
    }
  }

  onSelect = (date) => {
    const { input } = this.props;
    const newDate = new Date(date).toISOString();
    // update redux-form
    input.onChange(newDate);
    // update display field
    this.setState({ displayText: newDate.slice(0, 10) });
  }

  initPikaday = () => {
    const { input, minDate } = this.props;

    this.picker = new Pikaday({
      field: this.pickerRef.current,
      onSelect: this.onSelect,
      minDate,
      yearRange: [new Date().getFullYear(), new Date().getFullYear() + 30],
      trigger: this.displayRef.current,
    });

    if (input.value) {
      this.picker.setDate(input.value);
    }
  }

  render() {
    const {
      input, minDate, disabled, ...props
    } = this.props;
    const { displayText } = this.state;

    return (
      <div className="fieldDate has-feedback">
        {/* field that will be displayed */}
        <input
          ref={this.displayRef}
          className="form-control"
          disabled={disabled}
          value={displayText}
          onChange={() => {}}
        />
        {!disabled && <span className="ca-gov-icon-calendar form-control-feedback" />}
        {/* date picker */}
        <input
          ref={this.pickerRef}
          {...props}
          className="hidden"
        />
        {/* handle dispatching to redux-form */}
        <input
          className="hidden"
          id={input.name}
          onChange={(v) => { console.log('changed: ', v); }}
        />
      </div>
    );
  }
}

export default DatePicker;
