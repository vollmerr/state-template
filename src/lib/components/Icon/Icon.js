import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

/**
 * The `Icon` is used to render a state template icon,
 * automatically prefixing all icons with the correct class names.
 * See http://template.webstandards.ca.gov/sample/icon-fonts.html
 * for the full list of avaiable icon names.
 */
const Icon = (props) => {
  const {
    className,
    name,
    srOnly,
    tag: Tag,
    ...rest
  } = props;

  const cn = classNames([
    className,
    `ca-gov-icon-${name}`,
  ]);

  return (
    <>
      <Tag data-test={'icon'} className={cn} aria-hidden {...rest} />
      {srOnly && <span className={'sr-only'}>{srOnly}</span>}
    </>
  );
};

Icon.propTypes = {
  /** Class names to attach to the top level div */
  className: T.string,

  /** Name of icon to use (auto prefixes with ca-gov-icon-) */
  name: T.string.isRequired,

  /** Screen Reader only text */
  srOnly: T.string,

  /** HTML tag to render as  */
  tag: T.node,
};

Icon.defaultProps = {
  className: null,
  srOnly: '',
  tag: 'span',
};

export default Icon;
