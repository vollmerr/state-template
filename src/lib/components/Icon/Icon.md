```jsx
  <Icon name={'logo'} />
```
### custom tag
Render the component using a custom component tag

```jsx
  <Icon name={'apps'} tag={'small'} />
```

### passed props
Other properties get passed down to the icon component, such as id, class names, etc

```jsx
  <Icon name={'apps'} id={'iconId'} className={'m-l-lg'} style={{ color: 'red' }} />
```
