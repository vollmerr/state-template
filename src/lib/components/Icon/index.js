import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

/**
 * The `Icon` is used to display a state template icon,
 * automatically prefixing all icons with the correct class names.
 * See http://template.webstandards.ca.gov/sample/icon-fonts.html
 * for the full list of avaiable icon names.
 */
const Icon = (props) => {
  const {
    name,
    className,
    tag: Tag,
    ...rest
  } = props;

  const cn = classNames([
    className,
    `ca-gov-icon-${name}`,
  ]);

  return <Tag className={cn} {...rest} />;
};

Icon.propTypes = {
  /** Name of icon to use (auto prefixes with ca-gov-icon-) */
  name: T.string.isRequired,
  /** Class names to attach to the top level div */
  className: T.string,
  /** HTML tag to use  */
  tag: T.node,
};

Icon.defaultProps = {
  className: null,
  tag: 'span',
};

export default Icon;
