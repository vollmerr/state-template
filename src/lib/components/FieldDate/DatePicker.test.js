import React from 'react';
import { shallow } from 'enzyme';

import Icon from '../Icon';
import DatePicker from './DatePicker';

const { testUtils } = global;

const props = {
  input: testUtils.getInputProp({ value: '2001-05-09' }),
  disabled: false,
  minDate: '1900-01-02',
};

let wrapper;
let instance;
describe('DatePicker', () => {
  beforeEach(() => {
    wrapper = shallow(<DatePicker {...props} />);
    instance = wrapper.instance();
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should format and update the displayed date when a new value is recieved', () => {
    const input = { ...props.input, value: new Date('2011-09-01T08:00:00.000Z') };
    instance.componentWillReceiveProps({ ...props, input });
    expect(wrapper.state('displayText')).toBe('2011-09-01');
  });

  it('should hide the calender icon when disabled', () => {
    wrapper.setProps({ disabled: true });
    expect(wrapper.find(Icon).length).toBe(0);
  });

  it('should handle invalid dates', () => {
    const date = instance.getDate('invalid date');
    expect(date).toBe('');
  });
});
