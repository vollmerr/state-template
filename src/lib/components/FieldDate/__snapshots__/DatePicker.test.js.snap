// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker should render correctly 1`] = `
<div
  className="field-date has-feedback"
>
  <input
    autoComplete="off"
    className="form-control"
    disabled={false}
    id="testName"
    name="testName"
    onChange={[Function]}
    value="2001-05-08"
  />
  <Icon
    className="form-control-feedback"
    name="calendar"
    tag="span"
  />
  <input
    className="hidden"
  />
</div>
`;
